Flipper oop refactor

sehyunc opened the pull request:
TODO:
* refactor assert statements to require statements with message?
* possibly add/remove instance variables

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,108 @@
+"""
+
+
+class Flip:
```
Let's call this `Flipper` so it matches the the Maker code

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,108 @@
+
+
+class Flip:
+    ADDRESS = "flip"
```
Let's instantiate `ADDRESS` in the `__init__` function - since we'll have one `Flipper` per Ilk type, we should probably make the address something like `f"flipper_{ilk}"`

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,108 @@
+    kicks = None
+    bids = {}
+
+    def __init__(self, vat, beg, ttl, tau, kicks, bids):
```
We're missing an `ilk` class attribute, we should include it as a parameter to the `__init__` function & instantiate it there

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,108 @@
+        self.kicks = kicks
+        self.bids = bids
+
+    def flipper_kick(flipper, vat, user_id, gal, tab, lot, bid, end):
```
For this and the rest of the functions, we should:
1. Remove the `flipper_` prefix
2. Remove `flipper`, `vat`, and any other function parameters no longer necessary in the OOP setup
  - Make the appropriate changes, like using `self.vat` instead of `vat` and `self` instead of `flipper`
  - Oh also add `self` as the first parameter everywhere
3. Rename `user_id` to `usr` to match the Maker code

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,108 @@
+
+        curr_bid = flipper["bids"][bid_id]
+
+        assert (
```
We should replace the `asserts` everywhere with `require`, which you can import from `dai_cadcad.util`

akirillo commented:
ooh also for future feature branches in the OOP refactor, can you make the new branch off of the `oop-refactor` one as opposed to `master`? just keeps the diff pretty lol

sehyunc commented:
* renamed flip->flipper
* set ADDRESS in init
* add ilk to init
* refactored functions to follow OOP refactor
* replaced assert->require

TODO:
* not sure about parameters line 35
* just one ilk or multiple ilks in init?

akirillo commented:
could we rebase this onto the `oop-refactor` branch? or if a rebase is too scary, at least merge in `oop-refactor` to absorb some parts of the diff

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,116 @@
+            "tab": tab,
+        }
+
+        self.vat.flux(self.ilk, "cat", "flipper_eth", lot)
```
hmm, if `self.ilk` is a rich object/dict, then we'd have to use it's ID in the call to `flux` here. if it's just an ilk ID, can we rename the variable to `ilk_id` everywhere appropriate

sehyunc commented:
@akirillo sorry the diffs for the recent two commits are messy, I rearranged the instance variables and added cat. lmk if these changes look good, and I'll merge it in

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,119 @@
+        self.kicks += 1
+        bid_id = self.kicks
+
+        self.bids[bid_id] = {
```
hmm I think we should make a `Bid` class instead of just using a dict, there's a struct defined for it in the [source code](https://github.com/makerdao/dss/blob/master/src/flip.sol#L58) and we've already adopted the convention of creating classes for structs that are defined in the source code (e.g. Ilk classes)

akirillo replied:
and we should also save the bid's id as a class member on itself for convenience

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,134 @@
+    lot = Wad(0)
+    guy = ""
+    tic = ""  # Date
+    end = ""  # Date
```
let's switch `tic` and `end` to be integers (they'll represent #s of timesteps)

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,134 @@
+    def tend(self, bid_id, usr, lot, bid, now):
+        """Places a tend bid on a Flipper auction."""
+
+        curr_bid = self.bids[bid_id]
```
switch bracket-indexing to dot-indexing

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,134 @@
+    kicks = None
+    ttl = None
+    tau = None
+    vat = None
```
this is super nitpicky but can we have thee default values be this:
```
beg = Wad(0)
bids = {}
cat = None
ilk_id = ""
kicks = 0
ttl = 0
tau = 0
vat = None
```
reason being that this will indicate the data types of each of these instance variables

akirillo commented on dai_cadcad/flip.py:
```diff
@@ -0,0 +1,134 @@
+    tab = Rad(0)
+    id = ""
+
+    def __init__(self, bid, lot, guy, tic, end, usr, gal, tab, id):
```
let's have the parameter `id` here be called `bid_id` for verboseness (you can keep the name of the instance variable itself as `id`)

akirillo reviewed (approved):
lgtm
